core/sql/optimizer/ItemExpr.cpp

```diff
@@ -14906,3 +14906,80 @@ NABoolean LOBoper::isCovered
   return FALSE;
 }
 
+// Compute the exprssion at compile time. Assume all operands are constants.
+// Return NULL if the computation fails and CmpCommon::diags() may be side-affected.
+ConstValue* ItemExpr::compute(CollHeap* heap)
+{
+  ValueIdList exprs;
+  exprs.insert(getValueId());
+
+  const NAType& dataType = getValueId().getType();
+
+  Lng32 decodedValueLen = dataType.getNominalSize() + dataType.getSQLnullHdrSize();
+
+  char staticDecodeBuf[200];
+  Lng32 staticDecodeBufLen = 200;
+
+  char* decodeBuf = staticDecodeBuf;
+  Lng32 decodeBufLen = staticDecodeBufLen;
+
+  // For character types, multiplying by 8 to deal with conversions between
+  // any two known character sets supported.  
+  Lng32 factor = (DFS2REC::isAnyCharacter(dataType.getFSDatatype())) ? 8 : 1;
+
+  if ( staticDecodeBufLen < decodedValueLen * factor) {
+    decodeBufLen = decodedValueLen * factor;
+    decodeBuf = new (STMTHEAP) char[decodeBufLen];
```

Comment:
No. null will not be needed since a binary value will be the result.

On Mon, Oct 19, 2015 at 1:04 PM, DaveBirdsall <notifications@github.com>
wrote:

> In core/sql/optimizer/ItemExpr.cpp
> <https://github.com/apache/incubator-trafodion/pull/132#discussion_r42404746>
> :
>
> > +
> > +  Lng32 decodedValueLen = dataType.getNominalSize() + dataType.getSQLnullHdrSize();
> > +
> > +  char staticDecodeBuf[200];
> > +  Lng32 staticDecodeBufLen = 200;
> > +
> > +  char* decodeBuf = staticDecodeBuf;
> > +  Lng32 decodeBufLen = staticDecodeBufLen;
> > +
> > +  // For character types, multiplying by 8 to deal with conversions between
> > +  // any two known character sets supported.
> > +  Lng32 factor = (DFS2REC::isAnyCharacter(dataType.getFSDatatype())) ? 8 : 1;
> > +
> > +  if ( staticDecodeBufLen < decodedValueLen * factor) {
> > +    decodeBufLen = decodedValueLen * factor;
> > +    decodeBuf = new (STMTHEAP) char[decodeBufLen];
>
> We don't care about trailing nulls, right? (Hence, no +1 after
> decodeBufLen)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-trafodion/pull/132/files#r42404746>.
>



-- 
Regards, --Qifan